node/silkworm/stagedsync/sync_loop.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+    Copyright 2021 The Silkworm Authors
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#include "sync_loop.hpp"
+
+#include <boost/format.hpp>
+
+#include <silkworm/common/asio_timer.hpp>
+#include <silkworm/common/log.hpp>
+#include <silkworm/db/stages.hpp>
+#include <silkworm/stagedsync/stagedsync.hpp>
+
+namespace silkworm::stagedsync {
+
+void SyncLoop::load_stages() {
+    stages_.push_back(std::make_unique<stagedsync::BlockHashes>(node_settings_));
+    stages_.push_back(std::make_unique<stagedsync::Senders>(node_settings_));
+    stages_.push_back(std::make_unique<stagedsync::Execution>(node_settings_));
+}
+
+void SyncLoop::work() {
+    static std::string log_prefix_fmt{"Stage %u/%u : %s"};
+
+    log::Trace() << "Synchronization loop started";
+
+    bool is_first_cycle{true};
+    std::unique_ptr<db::RWTxn> cycle_txn{nullptr};
+    mdbx::txn_managed external_txn;
+
+    StopWatch stop_watch;
+    Timer log_timer(
+        node_settings_->asio_context, node_settings_->sync_loop_log_interval_seconds * 1'000,
+        [&]() -> bool {
+            if (is_stopping()) {
+                return false;
+            }
+            std::string prefix = boost::str(boost::format(log_prefix_fmt) % (current_stage_ + 1) % stages_.size() %
```

Comment:
(Minor) Would be nice to extract into a function since it's repeated at line 94.

Reply:
Create a function to replace a one liner call to a boost function ?
I'm puzzled

Reply:
Yes, something like
```
std::string SyncLoop::format_prefix() const {
    static std::string log_prefix_fmt{"Stage %u/%u : %s"};
    return boost::str(boost::format(log_prefix_fmt) % (current_stage_ + 1) % stages_.size() %
                      stages_.at(current_stage_)->name());
}
```
It's a minor thing to remove code duplication between lines 50 & 94.

Reply:
Be my guest ... bd8a622